x/erc20/keeper/genesis.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/evmos/evmos/v16/x/erc20/types"
+	"github.com/evmos/evmos/v16/x/evm/statedb"
+)
+
+// SetGenesisTokenPairs stores in state the provided token pairs in genesis
+func (k Keeper) SetGenesisTokenPairs(ctx sdk.Context, pairs []types.TokenPair) error {
+	for i, pair := range pairs {
+		contractAddr := pair.GetERC20Contract()
+		coinMeta, err := k.getGenesisTokenPairMeta(ctx, pair, i)
+		if err != nil {
+			return fmt.Errorf("error while generating metadata for genesis pair denom %s. %w", pair.Denom, err)
+		}
+		stateDB := statedb.New(ctx, k.evmKeeper, statedb.TxConfig{})
+		code, err := k.generateContractCode(ctx, stateDB, coinMeta, contractAddr)
+		if err != nil {
+			return fmt.Errorf("error while getting contract code for genesis pair denom %s. %w", pair.Denom, err)
+		}
+		stateDB.SetCode(contractAddr, code)
+
+		if err := stateDB.Commit(); err != nil {
+			return fmt.Errorf("error while storing contract for genesis pair denom %s. %w", pair.Denom, err)
+		}
+
+		id := pair.GetID()
+		k.SetTokenPair(ctx, pair)
+		k.SetDenomMap(ctx, pair.Denom, id)
+		k.SetERC20Map(ctx, pair.GetERC20Contract(), id)
+	}
+
+	return nil
+}
+
+// getGenesisTokenPairMeta is a helper function to generate token pair metadata for the genesis token pairs
+func (k Keeper) getGenesisTokenPairMeta(ctx sdk.Context, pair types.TokenPair, index int) (banktypes.Metadata, error) {
+	displayName := fmt.Sprintf("genToken%d", index)
+	meta := banktypes.Metadata{
+		Description: fmt.Sprintf("Genesis token pair number %d", index),
+		DenomUnits: []*banktypes.DenomUnit{
+			{Denom: pair.Denom, Exponent: 0, Aliases: []string{fmt.Sprintf("uGenToken%d", index)}},
+			{Denom: displayName, Exponent: 6},
+		},
+		Base:    pair.Denom,
+		Display: displayName,
+		Name:    displayName,
+		Symbol:  displayName,
+	}
+	if err := k.verifyMetadata(ctx, meta); err != nil {
+		return banktypes.Metadata{}, errorsmod.Wrapf(
+			types.ErrInternalTokenPair, "coin metadata is invalid for genesis pair denom %s", pair.Denom,
+		)
+	}
+	return meta, nil
+}
+
+// generateContractCode is a helper function to generate
+// the ERC20 contract code to be stored at genesis when token pairs are provided
+func (k Keeper) generateContractCode(ctx sdk.Context, stateDB *statedb.StateDB, coinMeta banktypes.Metadata, contractAddr common.Address) ([]byte, error) {
+	evm := k.newEVM(ctx, stateDB)
+	data, err := getContractDataBz(coinMeta)
+	if err != nil {
+		return nil, err
+	}
+	// Initialize a new contract and set the code that is to be used by the EVM.
+	// The contract is a scoped environment for this execution context only.
+	sender := vm.AccountRef(types.ModuleAddress)
+	contract := vm.NewContract(sender, vm.AccountRef(contractAddr), big.NewInt(0), 1000000)
+	contract.Code = data
+
+	return evm.Interpreter().Run(contract, nil, false)
```

Comment:
why are we calling `Run`?

Reply:
This generates the contract code to be stored in the contract address ([geth does it here](https://github.com/ethereum/go-ethereum/blob/2365d7796854744e8ba185dda855357e8fb9c292/core/vm/evm.go#L467-L486)). It runs the constructor and returns the resulting code to store.